File: ScanProject/JavaAnalysis/SonarJava/src/main/java/Hard/1945-longest-duplicate-substring.java
package main.java.Hard;

import java.util.HashSet;
import java.util.Set;

/**https://leetcode.com/problems/longest-duplicate-substring/ */
//Given a string s, consider all duplicated substrings: (contiguous) substrings of s that occur 2 or more times. The occurrences may overlap.
//Return any duplicated substring that has the longest possible length. If s does not have a duplicated substring, the answer is "".
// 
//Example 1:
//Input: s = "banana"
//Output: "ana"
//Example 2:
//Input: s = "abcd"
//Output: ""
// 
//Constraints:
//	2 <= s.length <= 3 * 104
//	s consists of lowercase English letters.
class Solution37 {
    public String longestDupSubstring(String s) {
        int n = s.length();
        int[] nums = new int[n];
        for(int i = 0; i < n; i++){
            nums[i] = s.charAt(i) - 'a';
        }
        int base = 26;
        long mod = (long) Math.pow(2, 32);
        int left = 1;
        int right = n;
        while(left < right){
            int mid = left + (right - left) / 2;
            if(search(mid, base, mod, n, nums) != -1){
                left = mid + 1;
            }else{
                right = mid;
            }
        }
        int start = search(left - 1, base, mod, n, nums);
        return start == -1 ? "" : s.substring(start, start + left - 1);
    }
    private int search(int mid, int base, long mod, int n, int[] nums){
        long hash = 0;
        for(int i = 0; i < mid; i++){
            hash = (hash * base + nums[i]) % mod;
        }
        Set<Long> set = new HashSet<>();
        set.add(hash);
        long mult = 1;
        for(int i = 0; i < mid; i++){
            mult = (mult * base) % mod;
        }
        for(int i = 1; i < n - mid + 1; i++){
            hash = (hash * base - nums[i - 1] * mult % mod + mod) % mod;
            hash = (hash + nums[i + mid - 1]) % mod;
            if(set.contains(hash)){
                return i;
            }
            set.add(hash);
        }
        return -1;
    }
}

File: ScanProject/JSAnalysis/SonarJS/Hard/1945-longest-duplicate-substring.js
/**https://leetcode.com/problems/longest-duplicate-substring/ */
//Given a string s, consider all duplicated substrings: (contiguous) substrings of s that occur 2 or more times. The occurrences may overlap.
//Return any duplicated substring that has the longest possible length. If s does not have a duplicated substring, the answer is "".


/**
 * @param {string} s
 * @return {string}
 */
var longestDupSubstring = function(s) {
    let n = s.length;
    let a = 26;
    let mod = 2**32;
    let left = 1;
    let right = n;
    let ans = '';
    while(left <= right){
        let mid = Math.floor((left + right) / 2);
        let found = rabinKarp(s, mid, a, mod);
        if(found != -1){
            ans = s.substring(found, found + mid);
            left = mid + 1;
        }
        else{
            right = mid - 1;
        }
    }
    return ans;
    
    
};
function rabinKarp(s, len, a, mod){
    let h = 0;
    let aL = 1;
    for(let i = 0; i < len; i++){
        h = (h * a + s.charCodeAt(i)) % mod;
        aL = (aL * a) % mod;
    }
    let seen = new Set();
    seen.add(h);
    for(let i = 1; i <= s.length - len; i++){
        h = (h * a - s.charCodeAt(i - 1) * aL % mod + mod) % mod;
        h = (h + s.charCodeAt(i + len - 1)) % mod;
        if(seen.has(h)){
            return i;
        }
        seen.add(h);
    }
    return -1;
}

    